File: 프로그래머스/2/42839. 소수 찾기/소수 찾기.py
from itertools import permutations
from math import sqrt

def solution(numbers):
    is_prime = [True] * 10**len(numbers)
    is_prime[0] = False
    is_prime[1] = False
    
    for i in range(2, int(sqrt(len(is_prime))) + 1):
        if not is_prime[i]: continue
        
        for j in range(i**2, len(is_prime), i):
            is_prime[j] = False
    
    answer = set()
    
    for length in range(1, len(numbers) + 1):
        for num_tuple in permutations(numbers, length):
            num = int("".join(num_tuple))
            if is_prime[num]: answer.add(num)
    
    return len(answer)

File: 프로그래머스/2/42839. 소수 찾기/소수 찾기.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <cmath>
#include <set>

using namespace std;

int solution(string numbers) {
    set<int> answer;
    
    vector<bool> is_prime(pow(10, numbers.size()), true);
    is_prime[0] = false;
    is_prime[1] = false;
    for(int n=2; n<is_prime.size(); n++){
        if(!is_prime[n]) continue;
        for(int i = n*n; i<is_prime.size(); i+=n){
            is_prime[i] = false;
        }
    }
    
    sort(numbers.begin(), numbers.end());
    do{
        for(int len=1; len<=numbers.size(); len++){
            int num = stoi(numbers.substr(0, len));
            if(is_prime[num]) answer.insert(num);
        }
    }while(next_permutation(numbers.begin(), numbers.end()));
    
    return answer.size();
}